File: src/main/java/net/pterodactylus/sone/web/OptionsPage.java
/*
 * Sone - OptionsPage.java - Copyright © 2010–2016 David Roden
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package net.pterodactylus.sone.web;

import static net.pterodactylus.sone.utils.NumberParsers.parseInt;

import java.util.ArrayList;
import java.util.List;

import net.pterodactylus.sone.core.Preferences;
import net.pterodactylus.sone.data.Sone;
import net.pterodactylus.sone.data.Sone.ShowCustomAvatars;
import net.pterodactylus.sone.fcp.FcpInterface.FullAccessRequired;
import net.pterodactylus.sone.web.page.FreenetRequest;
import net.pterodactylus.util.template.Template;
import net.pterodactylus.util.template.TemplateContext;
import net.pterodactylus.util.web.Method;

/**
 * This page lets the user edit the options of the Sone plugin.
 *
 * @author <a href="mailto:bombe@pterodactylus.net">David ‘Bombe’ Roden</a>
 */
public class OptionsPage extends SoneTemplatePage {

	/**
	 * Creates a new options page.
	 *
	 * @param template
	 *            The template to render
	 * @param webInterface
	 *            The Sone web interface
	 */
	public OptionsPage(Template template, WebInterface webInterface) {
		super("options.html", template, "Page.Options.Title", webInterface, false);
	}

	//
	// TEMPLATEPAGE METHODS
	//

	/**
	 * {@inheritDoc}
	 */
	@Override
	protected void processTemplate(FreenetRequest request, TemplateContext templateContext) throws RedirectException {
		super.processTemplate(request, templateContext);
		Preferences preferences = webInterface.getCore().getPreferences();
		Sone currentSone = webInterface.getCurrentSone(request.getToadletContext(), false);
		if (request.getMethod() == Method.POST) {
			List<String> fieldErrors = new ArrayList<String>();
			if (currentSone != null) {
				boolean autoFollow = request.getHttpRequest().isPartSet("auto-follow");
				currentSone.getOptions().setAutoFollow(autoFollow);
				boolean enableSoneInsertNotifications = request.getHttpRequest().isPartSet("enable-sone-insert-notifications");
				currentSone.getOptions().setSoneInsertNotificationEnabled(enableSoneInsertNotifications);
				boolean showNotificationNewSones = request.getHttpRequest().isPartSet("show-notification-new-sones");
				currentSone.getOptions().setShowNewSoneNotifications(showNotificationNewSones);
				boolean showNotificationNewPosts = request.getHttpRequest().isPartSet("show-notification-new-posts");
				currentSone.getOptions().setShowNewPostNotifications(showNotificationNewPosts);
				boolean showNotificationNewReplies = request.getHttpRequest().isPartSet("show-notification-new-replies");
				currentSone.getOptions().setShowNewReplyNotifications(showNotificationNewReplies);
				String showCustomAvatars = request.getHttpRequest().getPartAsStringFailsafe("show-custom-avatars", 32);
				currentSone.getOptions().setShowCustomAvatars(ShowCustomAvatars.valueOf(showCustomAvatars));
				webInterface.getCore().touchConfiguration();
			}
			Integer insertionDelay = parseInt(request.getHttpRequest().getPartAsStringFailsafe("insertion-delay", 16), null);
			if (!preferences.validateInsertionDelay(insertionDelay)) {
				fieldErrors.add("insertion-delay");
			} else {
				preferences.setInsertionDelay(insertionDelay);
			}
			Integer postsPerPage = parseInt(request.getHttpRequest().getPartAsStringFailsafe("posts-per-page", 4), null);
			if (!preferences.validatePostsPerPage(postsPerPage)) {
				fieldErrors.add("posts-per-page");
			} else {
				preferences.setPostsPerPage(postsPerPage);
			}
			Integer imagesPerPage = parseInt(request.getHttpRequest().getPartAsStringFailsafe("images-per-page", 4), null);
			if (!preferences.validateImagesPerPage(imagesPerPage)) {
				fieldErrors.add("images-per-page");
			} else {
				preferences.setImagesPerPage(imagesPerPage);
			}
			Integer charactersPerPost = parseInt(request.getHttpRequest().getPartAsStringFailsafe("characters-per-post", 10), null);
			if (!preferences.validateCharactersPerPost(charactersPerPost)) {
				fieldErrors.add("characters-per-post");
			} else {
				preferences.setCharactersPerPost(charactersPerPost);
			}
			Integer postCutOffLength = parseInt(request.getHttpRequest().getPartAsStringFailsafe("post-cut-off-length", 10), null);
			if (!preferences.validatePostCutOffLength(postCutOffLength)) {
				fieldErrors.add("post-cut-off-length");
			} else {
				preferences.setPostCutOffLength(postCutOffLength);
			}
			boolean requireFullAccess = request.getHttpRequest().isPartSet("require-full-access");
			preferences.setRequireFullAccess(requireFullAccess);
			Integer positiveTrust = parseInt(request.getHttpRequest().getPartAsStringFailsafe("positive-trust", 3), null);
			if (!preferences.validatePositiveTrust(positiveTrust)) {
				fieldErrors.add("positive-trust");
			} else {
				preferences.setPositiveTrust(positiveTrust);
			}
			Integer negativeTrust = parseInt(request.getHttpRequest().getPartAsStringFailsafe("negative-trust", 4), null);
			if (!preferences.validateNegativeTrust(negativeTrust)) {
				fieldErrors.add("negative-trust");
			} else {
				preferences.setNegativeTrust(negativeTrust);
			}
			String trustComment = request.getHttpRequest().getPartAsStringFailsafe("trust-comment", 256);
			if (trustComment.trim().length() == 0) {
				trustComment = null;
			}
			preferences.setTrustComment(trustComment);
			boolean fcpInterfaceActive = request.getHttpRequest().isPartSet("fcp-interface-active");
			preferences.setFcpInterfaceActive(fcpInterfaceActive);
			Integer fcpFullAccessRequiredInteger = parseInt(request.getHttpRequest().getPartAsStringFailsafe("fcp-full-access-required", 1), preferences.getFcpFullAccessRequired().ordinal());
			FullAccessRequired fcpFullAccessRequired = FullAccessRequired.values()[fcpFullAccessRequiredInteger];
			preferences.setFcpFullAccessRequired(fcpFullAccessRequired);
			webInterface.getCore().touchConfiguration();
			if (fieldErrors.isEmpty()) {
				throw new RedirectException(getPath());
			}
			templateContext.set("fieldErrors", fieldErrors);
		}
		if (currentSone != null) {
			templateContext.set("auto-follow", currentSone.getOptions().isAutoFollow());
			templateContext.set("enable-sone-insert-notifications", currentSone.getOptions().isSoneInsertNotificationEnabled());
			templateContext.set("show-notification-new-sones", currentSone.getOptions().isShowNewSoneNotifications());
			templateContext.set("show-notification-new-posts", currentSone.getOptions().isShowNewPostNotifications());
			templateContext.set("show-notification-new-replies", currentSone.getOptions().isShowNewReplyNotifications());
			templateContext.set("show-custom-avatars", currentSone.getOptions().getShowCustomAvatars().name());
		}
		templateContext.set("insertion-delay", preferences.getInsertionDelay());
		templateContext.set("posts-per-page", preferences.getPostsPerPage());
		templateContext.set("images-per-page", preferences.getImagesPerPage());
		templateContext.set("characters-per-post", preferences.getCharactersPerPost());
		templateContext.set("post-cut-off-length", preferences.getPostCutOffLength());
		templateContext.set("require-full-access", preferences.isRequireFullAccess());
		templateContext.set("positive-trust", preferences.getPositiveTrust());
		templateContext.set("negative-trust", preferences.getNegativeTrust());
		templateContext.set("trust-comment", preferences.getTrustComment());
		templateContext.set("fcp-interface-active", preferences.isFcpInterfaceActive());
		templateContext.set("fcp-full-access-required", preferences.getFcpFullAccessRequired().ordinal());
	}

}

File: src/main/kotlin/net/pterodactylus/sone/web/pages/OptionsPage.kt
package net.pterodactylus.sone.web.pages

import net.pterodactylus.sone.data.SoneOptions.*
import net.pterodactylus.sone.fcp.FcpInterface.*
import net.pterodactylus.sone.main.*
import net.pterodactylus.sone.utils.*
import net.pterodactylus.sone.web.*
import net.pterodactylus.sone.web.page.*
import net.pterodactylus.util.template.*
import javax.inject.*

/**
 * This page lets the user edit the options of the Sone plugin.
 */
@MenuName("Options")
@TemplatePath("/templates/options.html")
@ToadletPath("options.html")
class OptionsPage @Inject constructor(webInterface: WebInterface, loaders: Loaders, templateRenderer: TemplateRenderer) :
		SoneTemplatePage(webInterface, loaders, templateRenderer, pageTitleKey = "Page.Options.Title") {

	override fun handleRequest(soneRequest: SoneRequest, templateContext: TemplateContext) {
		if (soneRequest.isPOST) {
			val fieldsWithErrors = mutableListOf<String>()
			getCurrentSone(soneRequest.toadletContext)?.options?.let { options ->
				val autoFollow = "auto-follow" in soneRequest.parameters
				val loadLinkedImages = soneRequest.parameters["load-linked-images"].emptyToNull
				val showCustomAvatars = soneRequest.parameters["show-custom-avatars"].emptyToNull
				val enableSoneInsertNotification = "enable-sone-insert-notifications" in soneRequest.parameters
				val showNewSoneNotification = "show-notification-new-sones" in soneRequest.parameters
				val showNewPostNotification = "show-notification-new-posts" in soneRequest.parameters
				val showNewReplyNotification = "show-notification-new-replies" in soneRequest.parameters

				options.isAutoFollow = autoFollow
				options.isSoneInsertNotificationEnabled = enableSoneInsertNotification
				options.isShowNewSoneNotifications = showNewSoneNotification
				options.isShowNewPostNotifications = showNewPostNotification
				options.isShowNewReplyNotifications = showNewReplyNotification
				loadLinkedImages?.also { if (cantSetOption { options.loadLinkedImages = LoadExternalContent.valueOf(loadLinkedImages) }) fieldsWithErrors += "load-linked-images" }
				showCustomAvatars?.also { if (cantSetOption { options.showCustomAvatars = LoadExternalContent.valueOf(showCustomAvatars) }) fieldsWithErrors += "show-custom-avatars" }
			}
			val fullAccessRequired = "require-full-access" in soneRequest.parameters
			val fcpInterfaceActive = "fcp-interface-active" in soneRequest.parameters

			soneRequest.core.preferences.newRequireFullAccess = fullAccessRequired
			soneRequest.core.preferences.newFcpInterfaceActive = fcpInterfaceActive

			val postsPerPage = soneRequest.parameters["posts-per-page"]?.toIntOrNull()
			val charactersPerPost = soneRequest.parameters["characters-per-post"]?.toIntOrNull()
			val postCutOffLength = soneRequest.parameters["post-cut-off-length"]?.toIntOrNull()
			val imagesPerPage = soneRequest.parameters["images-per-page"]?.toIntOrNull()
			val insertionDelay = soneRequest.parameters["insertion-delay"]?.toIntOrNull()
			val fcpFullAccessRequired = soneRequest.parameters["fcp-full-access-required"]?.toIntOrNull()
			val negativeTrust = soneRequest.parameters["negative-trust"]?.toIntOrNull()
			val positiveTrust = soneRequest.parameters["positive-trust"]?.toIntOrNull()
			val trustComment = soneRequest.parameters["trust-comment"]?.emptyToNull

			if (cantSetOption { soneRequest.core.preferences.newPostsPerPage = postsPerPage }) fieldsWithErrors += "posts-per-page"
			if (cantSetOption { soneRequest.core.preferences.newCharactersPerPost = charactersPerPost }) fieldsWithErrors += "characters-per-post"
			if (cantSetOption { soneRequest.core.preferences.newPostCutOffLength = postCutOffLength }) fieldsWithErrors += "post-cut-off-length"
			if (cantSetOption { soneRequest.core.preferences.newImagesPerPage = imagesPerPage }) fieldsWithErrors += "images-per-page"
			if (cantSetOption { soneRequest.core.preferences.newInsertionDelay = insertionDelay }) fieldsWithErrors += "insertion-delay"
			fcpFullAccessRequired?.also { if (cantSetOption { soneRequest.core.preferences.newFcpFullAccessRequired = FullAccessRequired.values()[fcpFullAccessRequired] }) fieldsWithErrors += "fcp-full-access-required" }
			if (cantSetOption { soneRequest.core.preferences.newNegativeTrust = negativeTrust }) fieldsWithErrors += "negative-trust"
			if (cantSetOption { soneRequest.core.preferences.newPositiveTrust = positiveTrust }) fieldsWithErrors += "positive-trust"
			if (cantSetOption { soneRequest.core.preferences.newTrustComment = trustComment }) fieldsWithErrors += "trust-comment"

			if (fieldsWithErrors.isEmpty()) {
				soneRequest.core.touchConfiguration()
				throw RedirectException("options.html")
			}
			templateContext["fieldErrors"] = fieldsWithErrors
		}
		getCurrentSone(soneRequest.toadletContext)?.options?.let { options ->
			templateContext["auto-follow"] = options.isAutoFollow
			templateContext["show-notification-new-sones"] = options.isShowNewSoneNotifications
			templateContext["show-notification-new-posts"] = options.isShowNewPostNotifications
			templateContext["show-notification-new-replies"] = options.isShowNewReplyNotifications
			templateContext["enable-sone-insert-notifications"] = options.isSoneInsertNotificationEnabled
			templateContext["load-linked-images"] = options.loadLinkedImages.toString()
			templateContext["show-custom-avatars"] = options.showCustomAvatars.toString()
		}
		soneRequest.core.preferences.let { preferences ->
			templateContext["insertion-delay"] = preferences.insertionDelay
			templateContext["characters-per-post"] = preferences.charactersPerPost
			templateContext["fcp-full-access-required"] = preferences.fcpFullAccessRequired.ordinal
			templateContext["images-per-page"] = preferences.imagesPerPage
			templateContext["fcp-interface-active"] = preferences.fcpInterfaceActive
			templateContext["require-full-access"] = preferences.requireFullAccess
			templateContext["negative-trust"] = preferences.negativeTrust
			templateContext["positive-trust"] = preferences.positiveTrust
			templateContext["post-cut-off-length"] = preferences.postCutOffLength
			templateContext["posts-per-page"] = preferences.postsPerPage
			templateContext["trust-comment"] = preferences.trustComment
		}
	}

	private fun cantSetOption(setter: () -> Unit) =
			try {
				setter()
				false
			} catch (iae: IllegalArgumentException) {
				true
			}

}
